src/cpp/fastdds/domain/DomainParticipantFactory.cpp

```diff
@@ -0,0 +1,239 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file DomainParticipantFactory.cpp
+ *
+ */
+
+#include <fastdds/domain/DomainParticipantFactory.hpp>
+#include <fastrtps/rtps/RTPSDomain.h>
+
+#include <fastdds/domain/DomainParticipant.hpp>
+#include "DomainParticipantImpl.hpp"
+
+#include <fastrtps/utils/eClock.h>
+#include <fastrtps/log/Log.h>
+
+#include <fastrtps/xmlparser/XMLProfileManager.h>
+
+#include <fastrtps/types/DynamicTypeBuilderFactory.h>
+#include <fastrtps/types/DynamicDataFactory.h>
+#include <fastrtps/types/TypeObjectFactory.h>
+
+using namespace eprosima::fastrtps::xmlparser;
+
+using eprosima::fastrtps::ParticipantAttributes;
+using eprosima::fastrtps::Log;
+
+using eprosima::fastrtps::rtps::RTPSDomain;
+using eprosima::fastrtps::rtps::RTPSParticipant;
+
+namespace eprosima {
+namespace fastdds {
+
+class DomainParticipantFactoryReleaser
+{
+public:
+    ~DomainParticipantFactoryReleaser()
+    {
+        DomainParticipantFactory::delete_instance();
+    }
+};
+
+static DomainParticipantFactoryReleaser s_releaser;
+static DomainParticipantFactory* g_instance = nullptr;
+
+DomainParticipantFactory::DomainParticipantFactory()
+    : default_xml_profiles_loaded(false)
+{
+}
+
+DomainParticipantFactory::~DomainParticipantFactory()
+{
+    {
+        std::lock_guard<std::mutex> guard(mtx_participants_);
+        for (auto it : participants_)
+        {
+            for (auto pit : vit.second)
+            {
+                pit->disable();
+            }
+        }
+        // Delete participants
+        for (auto vit : participants_)
+        {
+            for (auto pit : vit.second)
+            {
+                delete pit;
+            }
+        }
```

Comment:
It can produce invalid reads due to receiving events from the RTPS layer on other participants and endpoints destruction in a "destruction in progress" participant.

Reply:
:+1: